theano/gpuarray/dnn.py

```diff
@@ -2738,6 +2738,305 @@ def dnn_batch_normalization_test(inputs, gamma, beta, mean, var,
+class GpuDnnTransformerGradT(DnnBase):
+    """
+    Gradient of affine transformations Op for cuDNN Spatial Transformer.
+    """
+    __props__ = ()
+    _cop_num_inputs = 2
+    _cop_num_outputs = 1
+    _f16_ok = True
+    check_input = False
+
+    def __init__(self):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_gt.c"], "APPLY_SPECIFIC(dnn_sptf_gt)")
+
+    def make_node(self, dgrid, desc):
+        context_name = infer_context_name(desc)
+
+        if (not isinstance(desc.type, CDataType) or
+                desc.type.ctype != 'cudnnSpatialTransformerDescriptor_t'):
+            raise ValueError('desc must be cudnnSpatialTransformerDescriptor_t')
+
+        dgrid = as_gpuarray_variable(dgrid, context_name)
+        assert dgrid.dtype in ('float16', 'float32', 'float64')
+        assert dgrid.ndim == 4
+
+        dtheta = GpuArrayType(dtype=dgrid.dtype,
+                              broadcastable=(dgrid.type.ndim - 1) * (False,),
+                              context_name=context_name)()
+        inputs = [dgrid, desc]
+        outputs = [dtheta]
+
+        return Apply(self, inputs, outputs)
+
+    def connection_pattern(self, node):
+        # not connected to desc
+        return [[1], [0]]
+
+
+def dnn_spatialtf(img, theta, scale_width=1, scale_height=1, precision=theano.config.floatX):
+    """
+    GPU spatial transformer using cuDNN from NVIDIA.
+
+    Parameters
+    ----------
+    img : tensor
+        Images to which the transformations will be applied. The implementation
+        assumes the tensor is in NCHW format, where N is the number of images,
+        C is the number of color channels, H is the height of the inputs, and
+        W is width of the inputs.
+    theta : tensor
+        Affine transformation tensor containing one affine transformation
+        matrix per image. ``theta`` is usually generated by the localization
+        network.
+    scale_height: float
+        A float specifying the scaling factor for the height of the output
+        image. A value of 1 will keep the original height of the input. Values
+        larger than 1 will upsample the input. Values below 1 will downsample
+        the input.
+    scale_width: float
+        A float specifying the scaling factor for the width of the output
+        image. A value of 1 will keep the original width of the input. Values
+        larger than 1 will upsample the input. Values below 1 will downsample
+        the input.
+
+    Returns
+    -------
+    out : tensor
+        Transformed images with width and height properly scaled.
+
+    Notes
+    -----
+    Currently, cuDNN only supports 2D transformations with 2x3 affine
+    transformation matrices.
+
+    Also, the only grid sampler method available is the bilinear interpolation.
+    """
+    out_dims = (img.shape[0], img.shape[1],
+                img.shape[2] * scale_height,
+                img.shape[3] * scale_width)
+    out_dims = tuple([as_scalar(v).astype('int64') for v in out_dims])
```

Comment:
I've kept it this way because when img is a Variable, trying to create a NumPy array from ``out_dims`` fails, because the shape of the Variable is not known in advance.